### PR TITLE
Enable async service instance provisioning

### PR DESCRIPTION
**Pull request by samedguener**
- PR as discussed in issue #51
- This solves also issue #40

### CODE REVIEW

File: website/docs/r/service_instance.html.markdown [hunk 1]
```diff
@@ -35,7 +35,9 @@ The following arguments are supported:
 * `space` - (Required, String) The ID of the [space](/docs/providers/cloudfoundry/r/space.html) 
 * `json_params` - (Optional, String) Json string of arbitrary parameters. Some services support providing additional configuration parameters within the provision request
 * `tags` - (Optional, List) List of instance tags. Some services provide a list of tags that Cloud Foundry delivers in [VCAP_SERVICES Env variables](https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#VCAP-SERVICES)
-
+* `timeout` - (Optional, Integer) The timeout in seconds, which shall be waited until service instance is created. Default: 60 seconds
+* `async` - (Optional, Bool) 	Set to `true` if the client allows asynchronous provisioning. Please refer to [this](https://github.com/openservicebrokerapi/servicebroker/blob/v2.12/spec.md#synchronous-and-asynchronous-operations) for more information. Default: false
```
Comment (gberche-orange) [inline website/docs/r/service_instance.html.markdown]: Is this flag really necessary ? 
The provider should systematically set the `accepts_incomplete` flag in the [CC API call](http://apidocs.cloudfoundry.org/280/service_instances/creating_a_service_instance.html) and inter from the response status code (202 vs 200) and/or `last_operation.state` response field whether the service instance completion is sync or async

Comment (doktorgibson) [reply website/docs/r/service_instance.html.markdown]: Oh I see, the best way is to catch up if the given response is 200 or 202. I'll change the PR and according our the UX recommendations as discussed in [#51](https://github.com/mevansam/terraform-provider-cf/issues/51) beginning of next week.

 Thanks for the review!

**Comment by gberche-orange** [comment]
@samedguener also we probably also need to add async service test coverage into https://github.com/mevansam/terraform-provider-cf/blob/master/cloudfoundry/resource_cf_service_instance_test.go. Potentially, we'd need to deploy a service broker which supports async provisionning, may be https://github.com/cloudfoundry/cf-acceptance-tests/blob/master/assets/service_broker/README.md can be useful there.

**Comment by doktorgibson** [comment]
@gberche-orange I have updated the PR according to our needs. Can you assist me on how we could enable async. provisioning in the service broker? I couldn't figure out, how and where the service broker is actually set-up during testing.

Thanks in advance for the clarification!

**Comment by gberche-orange** [comment]
@doktorgibson the documentation at https://github.com/cloudfoundry/cf-acceptance-tests/blob/master/assets/service_broker/README.md#the-configuration-file describes how the configuration time allows to specify responses to each endpoint (status code and body). E.g. 

```json
   "async-plan-guid": {
      "sleep_seconds": 0,
      "status": 202,
      "body": {
        "last_operation": {
          "state": "in progress"
        }
      }

```
you might also get some inspiration from the related cf-acceptance-test async test cases at https://github.com/orange-cloudfoundry/cf-ops-automation/issues/62

**Comment by gberche-orange** [comment]
@doktorgibson Fixed the pointer to acceptance test, sorry for copy/paste error.

**Comment by gberche-orange** [comment]
@samedguener thanks for the test case contribution! Might be easier to maintain if we submodule the acceptance test repo instead of duplicating/vendoring the app.

**Comment by samedguener** [comment]
@gberche-orange This is not the final solution, there are still some few things missing. 

I will add it and notify you, if the PR is able to merge :)

**Comment by samedguener** [comment]
Hi @gberche-orange ,

I moved the service broker to ``` vendor/ ``` but we have to edit some configuration files in the service broker. As I moved it to ``` vendor/ ``` it became a submodule. How can we push our custom config for the service broker to our repository rather than pushing to upstream?

I have implemented all tests, there are still one problem with the service broker. Unfortunately I did not have the time to have a look into this. Can you take over for this PR and make the needed changes?

Best,
Samed

**Comment by gberche-orange** [comment]
thanks @samedguener 

> How can we push our custom config for the service broker to our repository rather than pushing to upstream of cf-acceptance-tests?

Pushing custom config needs to be performed using a curl command as mentioned in [broker config doc](https://github.com/cloudfoundry/cf-acceptance-tests/blob/master/assets/service_broker/README.md#the-configuration-file)

`curl <app_url>/config -X POST -d @data.json`

This may be invoked using a [local-exec provisionner](https://www.terraform.io/docs/provisioners/local-exec.html) part of the `+resource "cf_app" "fake-service-broker"` (I assume curl binary is available in the acceptance testing environment in travis-ci)

https://github.com/chrismathias/terraform-provider-cf/blob/e6aa34cd9e6c9c560f3c599089d0e19be3f220d9/cloudfoundry/resource_cf_service_instance_test.go#L82-L92

> I have implemented all tests, there are still one problem with the service broker. Unfortunately I did not have the time to have a look into this. Can you take over for this PR and make the needed changes?

The team is focusing on 1.0 milestone, but can support you or your colleagues to finish up the PR. Otherwise, we'll resume work on this support once 1.0 is released.

Also the PR needs to be rebased against the `dev` branch, see [CONTRIBUTING.md](https://github.com/mevansam/terraform-provider-cf/blob/master/CONTRIBUTING.md)

**Comment by gberche-orange** [comment]
@samedguener in response to your slack private message

> Hi, I am having following issue by trying to push the custom config using local-exec in the test :
```testing.go:513: Step 0 error: config is invalid: cf_service_broker.fake-service-broker: provisioner local-exec couldn't be found``` 
> It seems like the provisioner is not available in the test environment, I searched for that in our other tests, but did not find anything similar. I would be pleased if you have an idea on this.

Maybe the AWS provider can provide some inspiration to import the local-exec provisionner: 
- [external providers are vendored](https://github.com/terraform-providers/terraform-provider-aws/tree/master/vendor/github.com/terraform-providers), e.g. TLS
- it makes uses of the TLS provider in its tests by [importing the provider code in the test](https://github.com/terraform-providers/terraform-provider-aws/blob/b0287801870e0b4bbf4f2e0cd9a6ddfb49b298ab/aws/resource_aws_load_balancer_backend_server_policy_test.go#L13-L24) and then [using a TLS resource](https://github.com/terraform-providers/terraform-provider-aws/blob/b0287801870e0b4bbf4f2e0cd9a6ddfb49b298ab/aws/resource_aws_load_balancer_backend_server_policy_test.go#L148-L164) in the test case

Alternatively, the `data.json` file could be modified before pushing it with the broker app to CF, avoiding the need for runtime configuration through the provisioner ?

@ArthurHlt @mevansam @psycofdj any other simpler suggestion to configure the broker to be async ?

**Comment by jcarrothers-sap** [comment]
I just wanted to chime in with some feedback.  I am currently working on improving the provider myself for the needs of my own team and I actually need the async service instance creation/update feature internally right now.  However, I really cannot use this change as it is because it includes a massive set of changes that are purely cleanup in nature and are causing massive conflicts with my own changes.  

It would be really nice if this change could be parred down to just the implementation of that stated goal in the PR description and to move the various cleanup and other code improvements to a different PR.

**Comment by bgandon** [comment]
I'm sorry to repeat what Guillaume said last month, but @samedguener, could you please rebase this work against the `dev` branch? Thank you.

**Comment by jcarrothers-sap** [comment]
@samedguener And would it be possible to re-target this PR at the dev branch?  That would make more sense than trying to merge all of dev into master as this PR is currently attempting to do.

**Comment by janosbinder** [comment]
Hi @gberche-orange ,

> Maybe the AWS provider can provide some inspiration to import the local-exec provisionner:
> external providers are vendored, e.g. TLS
> it makes uses of the TLS provider in its tests by importing the provider code in the test and then using a TLS resource in the test case

Thanks this is a very good point. The problem is that the `local-exec` is a provisioner (uses the  `terraform.ResourceProvisioner` interface), while the test you have cited uses `terraform.ResourceProvider` interface. Therefore I don't know how to call/add to the config a provisioner.

I have also searched the tests of terraform and some terraform providers 
(e.g. aws, azure, openstack or gcp) for hint, but I did not find any examples to use a provisioner in a test.

In summary I would like to finish Samed's tests, however I am stuck with this issue. Do you any idea how we can pass this json to the service broker app? (see: https://github.com/chrismathias/terraform-provider-cf/blob/d0072d35b0abe34dadbf2ffe7f2ebbbb9077b325/cloudfoundry/resource_cf_service_instance_test.go#L92-L94)

**Comment by gberche-orange** [comment]
@janosbinder 

> Do you any idea how we can pass this json to the service broker app? (see: https://github.com/chrismathias/terraform-provider-cf/blob/d0072d35b0abe34dadbf2ffe7f2ebbbb9077b325/cloudfoundry/resource_cf_service_instance_test.go#L92-L94)

The test code may generate a config.json which overrides the service_broker default config file, prior triggering the cf-push through the cf_app resource.

See https://github.com/cloudfoundry/cf-acceptance-tests/tree/master/assets/service_broker#bootstrapping

```
The repo also includes a ruby script setup_new_broker.rb to bootstrap a new broker. 
The script does the following:
-    generates a unique config for the broker in order avoid guid and name conflicts with other brokers.
-    pushes the broker as an app
```

**Comment by janosbinder** [comment]
So I have extended service instance deletion to support asynchronous operations. When I run the tests I get the following error:

> --- FAIL: TestAccServiceBroker_async (174.55s)
> 	testing.go:513: Step 0 error: After applying this step and refreshing, the plan was not empty:
> 
> 		DIFF:
> 
> 		CREATE: data.cf_service.fake-service
> 		  name:            "" => "fake-service"
> 		  service_plans.%: "" => "<computed>"
> 
> 		STATE:
> 
> 		cf_app.fake-service-broker:
> 		  ID = 82c521ba-3d1e-4968-86eb-f744e21ab758
> 		  provider = provider.cf
> 		  disk_quota = 1024
> 		  enable_ssh = true
> 		  environment.% = 0
> 		  health_check_type = port
> 		  instances = 1
> 		  memory = 1024
> 		  name = fake-service-broker
> 		  ports.# = 1
> 		  ports.8080 = 8080
> 		  route.# = 1
> 		  route.0.default_route = 879284df-bd43-4e0a-8049-f134afe7e01c
> 		  route.0.default_route_mapping_id = 0045738b-bfd2-4cdb-9368-6e944edfff84
> 		  route.0.live_route =
> 		  route.0.live_route_mapping_id =
> 		  route.0.stage_route =
> 		  route.0.stage_route_mapping_id =
> 		  route.0.validation_script =
> 		  route.0.version =
> 		  space = 92f6f86b-9a89-4115-8bc4-bf2a349ea208
> 		  stack = 4be4dd1b-d234-45b3-8c0d-f136dd5d4ea0
> 		  stopped = false
> 		  timeout = 700
> 		  url = file://../vendor/github.com/cf-acceptance-tests/assets/service_broker/
> 
> 		  Dependencies:
> 		    cf_route.fake-service-broker-route
> 		    data.cf_space.space
> 		cf_route.fake-service-broker-route:
> 		  ID = 879284df-bd43-4e0a-8049-f134afe7e01c
> 		  provider = provider.cf
> 		  domain = 094e5d51-7e2b-47be-932b-f6761a2e3f98
> 		  endpoint = fake-service-broker.apps.pcf.cf1.tfacc.pcfs.io
> 		  hostname = fake-service-broker
> 		  path =
> 		  space = 92f6f86b-9a89-4115-8bc4-bf2a349ea208
> 
> 		  Dependencies:
> 		    data.cf_domain.fake-service-broker-domain
> 		    data.cf_space.space
> 		cf_service_broker.fake-service-broker:
> 		  ID = 9b9b27fe-3999-4d08-8046-73adcd048b84
> 		  provider = provider.cf
> 		  name = fake-service-broker
> 		  password = admin
> 		  service_plans.% = 3
> 		  service_plans.fake-service/fake-async-only-plan = 52e2d50c-589b-4e5e-8d22-785d457c9b6d
> 		  service_plans.fake-service/fake-async-plan = ae4a0975-7551-4149-b14d-bd56dff81d6e
> 		  service_plans.fake-service/fake-plan = ec5dbd02-8fc0-49e2-b701-3d5e6bc84d21
> 		  space = 92f6f86b-9a89-4115-8bc4-bf2a349ea208
> 		  url = http://fake-service-broker.apps.pcf.cf1.tfacc.pcfs.io
> 		  username = admin
> 
> 		  Dependencies:
> 		    cf_app.fake-service-broker
> 		    data.cf_space.space
> 		cf_service_instance.fake-service:
> 		  ID = 69eea449-ddb3-40e4-8611-2fb28b2b47a3
> 		  provider = provider.cf
> 		  name = fake-service
> 		  service_plan = ae4a0975-7551-4149-b14d-bd56dff81d6e
> 		  space = 92f6f86b-9a89-4115-8bc4-bf2a349ea208
> 		  tags.# = 0
> 
> 		  Dependencies:
> 		    cf_app.fake-service-broker
> 		    cf_service_broker.fake-service-broker
> 		    data.cf_space.space
> 		data.cf_domain.fake-service-broker-domain:
> 		  ID = 094e5d51-7e2b-47be-932b-f6761a2e3f98
> 		  provider = provider.cf
> 		  domain = pcf.cf1.tfacc.pcfs.io
> 		  name = apps.pcf.cf1.tfacc.pcfs.io
> 		  org =
> 		  sub_domain = apps
> 		data.cf_org.org:
> 		  ID = 0c080a9e-c935-4d76-a0ba-b9ffe4dcbd4e
> 		  provider = provider.cf
> 		  name = pcfdev-org
> 		data.cf_space.space:
> 		  ID = 92f6f86b-9a89-4115-8bc4-bf2a349ea208
> 		  provider = provider.cf
> 		  name = pcfdev-space
> 		  org = 0c080a9e-c935-4d76-a0ba-b9ffe4dcbd4e
> 		  org_name = pcfdev-org
> 		  quota =
> 
> 		  Dependencies:
> 		    data.cf_org.org

Any idea, how we could fix tests like this?

File: cloudfoundry/cfapi/service_manager.go [hunk 2]
```diff
@@ -505,6 +507,97 @@ func (sm *ServiceManager) ReadServiceInstance(serviceInstanceID string) (service
 	return serviceInstance, nil
 }
 
+// WaitServiceInstanceTo -
+func (sm *ServiceManager) WaitServiceInstanceTo(operationType string, serviceInstanceID string) (err error) {
+	sm.log.UI.Say("Waiting for service instance %s to finish starting ..", terminal.EntityNameColor(serviceInstanceID))
```
Comment (gberche-orange) [inline cloudfoundry/cfapi/service_manager.go]: Alternative wording suggestion: "Waiting for service instance %s to complete async provisionning"

File: cloudfoundry/cfapi/service_manager.go [hunk 3]
```diff
@@ -505,6 +507,97 @@ func (sm *ServiceManager) ReadServiceInstance(serviceInstanceID string) (service
...
+func (sm *ServiceManager) WaitDeletionServiceInstance(serviceInstanceID string) (err error) {
+	sm.log.UI.Say("Waiting for service instance %s to delete ..", terminal.EntityNameColor(serviceInstanceID))
+	c := make(chan error)
+	go func() {
+
+		var err error
+		var serviceInstance CCServiceInstance
+
+		for {
+			time.Sleep(appStatePingSleep)
```
Comment (gberche-orange) [inline cloudfoundry/cfapi/service_manager.go]: Same, please  consider leveraging the [StateChangeConf](https://github.com/terraform-providers/terraform-provider-aws/blob/9fc15be7dd071e12e1e1468e998a070ed197d385/aws/resource_aws_db_instance.go#L527-L545) helper  using the same polling loop pattern than what the aws_db_instance does

File: cloudfoundry/cfapi/service_manager.go [hunk 4]
```diff
@@ -505,6 +507,97 @@ func (sm *ServiceManager) ReadServiceInstance(serviceInstanceID string) (service
...
+			if serviceInstance, err = sm.ReadServiceInstance(serviceInstanceID); err != nil {
+				// if the service instance is gone the error message should contain 60004
+				// cf_service_instance.redis: Server error, status code: 404, error code: 60004, message: The service instance could not be found: babababa-d977-4e9c-9bd0-4903d146d822
+				if strings.Contains(err.Error(), "error code: 60004") {
```
Comment (gberche-orange) [inline cloudfoundry/cfapi/service_manager.go]: Minor alternative wording suggestion

// if the service instance is gone the error message should contain error code 60004 ("ServiceInstanceNotFound")
+				// e.g. CLI output: cf_service_instance.redis: Server error, status code: 404, error code: 60004, message: The service instance could not be found: babababa-d977-4e9c-9bd0-4903d146d822

File: cloudfoundry/resource_cf_service_instance.go [hunk 5]
```diff
@@ -157,26 +168,70 @@ func resourceServiceInstanceUpdate(d *schema.ResourceData, meta interface{}) (er
...
 	sm := session.ServiceManager()
 
-	err = sm.DeleteServiceInstance(d.Id())
-	if err != nil {
+	if err = sm.DeleteServiceInstance(id); err != nil {
+		return err
+	}
+
+	// Check whether service_instance has been indeed deleted (sometimes takes very long))
+	if err = sm.WaitDeletionServiceInstance(id); err != nil {
```
Comment (gberche-orange) [inline cloudfoundry/resource_cf_service_instance.go]: Althrough the [CC API doc for service instance deletion](http://apidocs.cloudfoundry.org/1.40.0/service_instances/delete_a_service_instance.html) is'nt very explicit, the async deletion should return a 202 whereas a sync deletion should a 204, see [associated CC specs](https://github.com/cloudfoundry/cloud_controller_ng/blob/6fac1e18f93599c24ec471ce4d7ec4f9c1d95e9c/spec/unit/controllers/services/service_instances_controller_spec.rb#L2421-L2433)

In the case of a 204 return status, the service status polling is unnecessary

File: cloudfoundry/resource_cf_service_instance.go [hunk 6]
```diff
@@ -157,26 +168,70 @@ func resourceServiceInstanceUpdate(d *schema.ResourceData, meta interface{}) (er
...
+
+	sm := session.ServiceManager()
+
+	serviceinstance, err := sm.ReadServiceInstance(d.Id())
+
+	if err != nil {
+		return nil, err
+	}
+
+	d.Set("name", serviceinstance.Name)
```
Comment (gberche-orange) [inline cloudfoundry/resource_cf_service_instance.go]: Most of the importers contributed by @ArthurHlt in https://github.com/mevansam/terraform-provider-cf/pull/14/files use the `ImportStatePassthrough` pattern. Would it also be applicable here ?

File: tests/data.json [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+      "fake-async-plan-guid": {
+        "sleep_seconds": 0,
+        "status": 202,
+        "body": {
+          "last_operation": {
+            "state": "in progress"
+          }
+        }
+      },
+      "fake-async-plan-2-guid": {
```
Comment (gberche-orange) [inline tests/data.json]: may be also cover the case of a service with async provisionning, and **sync** deprovisionning.

File: website/docs/r/app.html.markdown [hunk 8]
```diff
@@ -74,7 +74,9 @@ One of the following arguments must be declared to locate application source or
 
 * `service_binding` - (Optional, Array) Service instances to bind to the application.
 
-  - `service` - (Required, String) The service instance GUID.
+* `service_binding` - (Optional, Array) Service instances to bind to.
```
Comment (gberche-orange) [inline website/docs/r/app.html.markdown]: duplicated line ?

File: website/docs/r/service_instance.html.markdown [hunk 9]
```diff
@@ -33,8 +33,8 @@ The following arguments are supported:
 * `name` - (Required, String) The name of the Service Instance in Cloud Foundry
 * `service_plan` - (Required, String) The ID of the [service plan](/docs/providers/cloudfoundry/d/service_plan.html)
 * `space` - (Required, String) The ID of the [space](/docs/providers/cloudfoundry/r/space.html) 
-* `json_params` - (Optional, String) Json string of arbitrary parameters. Some services support providing additional configuration parameters within the provision request. By default, no params are provided.
-* `tags` - (Optional, List) List of instance tags. Some services provide a list of tags that Cloud Foundry delivers in [VCAP_SERVICES Env variables](https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#VCAP-SERVICES). By default, no tags are assigned.
+* `json_params` - (Optional, String) Json string of arbitrary parameters. Some services support providing additional configuration parameters within the provision request
```
Comment (gberche-orange) [inline website/docs/r/service_instance.html.markdown]: Needs a rebase of changes made into #73 that would otherwise be removed

File: website/docs/r/service_instance.html.markdown [hunk 10]
```diff
@@ -33,8 +33,8 @@ The following arguments are supported:
...
+* `tags` - (Optional, List) List of instance tags. Some services provide a list of tags that Cloud Foundry delivers in [VCAP_SERVICES Env variables](https://docs.cloudfoundry.org/devguide/deploy-apps/environment-variable.html#VCAP-SERVICES)
```
Comment (gberche-orange) [inline website/docs/r/service_instance.html.markdown]: Needs a rebase of changes made into #73 that would otherwise be removed

File: website/docs/r/service_instance.html.markdown [hunk 11]
```diff
@@ -49,3 +49,9 @@ An existing Service Instance can be imported using its guid, e.g.
 ```
 $ terraform import cf_service.redis a-guid
 ```
+
```
Comment (gberche-orange) [inline website/docs/r/service_instance.html.markdown]: Is this specific to the service_instance resource, or global timeouts that would apply on any resource ?

File: cloudfoundry/resource_cf_service_instance.go (hunk 5 above)
Comment (janosbinder) [reply cloudfoundry/resource_cf_service_instance.go]: This is a very good point. Unfortunately the status code is shallowed here:

https://github.com/mevansam/terraform-provider-cf/blob/3af2dbfd6f7f2477ebba5c1c6d8082bf2a252991/vendor/code.cloudfoundry.org/cli/cf/net/gateway.go#L178

But I come up with an alternative solution. First I try to delete the service instance via synchronous service operation, if that does not work, then asynchronous + pulling. See this commit for more details: https://github.com/mevansam/terraform-provider-cf/pull/52/commits/86a28396fea8e09d44af958b163a15b7389a2982

File: website/docs/r/service_instance.html.markdown (hunk 11 above)
Comment (janosbinder) [reply website/docs/r/service_instance.html.markdown]: It should be specific to the service_instance resource, and my recent commit enforces it: https://github.com/mevansam/terraform-provider-cf/pull/52/commits/72c9af8aa6d6e59fb8def1c342470e21e54d874c (See the StateChangeConf implementations)

File: cloudfoundry/resource_cf_service_instance.go (hunk 6 above)
Comment (janosbinder) [reply cloudfoundry/resource_cf_service_instance.go]: Thanks @gberche-orange . I have added that pattern, and fix the broken import, and extended the test for that.

File: tests/data.json (hunk 7 above)
Comment (janosbinder) [reply tests/data.json]: Unfortunately this test broke. The way I managed to fix it to only allowing asynchronous deprovisioning.

File: cloudfoundry/resource_cf_service_instance.go (hunk 5 above)
Comment (janosbinder) [reply cloudfoundry/resource_cf_service_instance.go]: I had to revert this change (https://github.com/mevansam/terraform-provider-cf/pull/52/commits/5f90fbfa5abdd2f2e92862b58c0739059a7c01f4). Unfortunately it broke  the test async provisioning, and sync deprovisioning. I tried to call this endpoint https://github.com/mevansam/terraform-provider-cf/blob/master/vendor/code.cloudfoundry.org/cli/cf/net/gateway.go#L127-L129 by rewriting this part https://github.com/chrismathias/terraform-provider-cf/blob/72c9af8aa6d6e59fb8def1c342470e21e54d874c/cloudfoundry/cfapi/service_manager.go#L555-L557 and trying the _accepts_incomplete=false_ and  _async=false_ flags, but it did not help. (see the flags here: https://apidocs.cloudfoundry.org/1.40.0/service_instances/delete_a_service_instance.html )

**Comment by janosbinder** [comment]
@gberche-orange thank you for your valuable comments. I am finished implementing them, namely:

File: website/docs/r/service_instance.html.markdown (hunk 11 above)
Comment (gberche-orange) [reply website/docs/r/service_instance.html.markdown]: thanks

File: tests/data.json (hunk 7 above)
Comment (gberche-orange) [reply tests/data.json]: thanks. This is a corner case we can probably live with for now. You may record the limitation into a distinct issue ?

Comment (janosbinder) [reply tests/data.json]: Limitation has been added https://github.com/mevansam/terraform-provider-cf/issues/107

**Comment by gberche-orange** [comment]
thanks @janosbinder for the contributed updates and detailed explanations.

> on deletion distinguishing between synchronous and asynchronous deprovisioning via HTTP status code - please see my comments in the review - briefly reciting them - the HTTP status code is shallowed by CF CLI gateway login, and my initial solution broke fake-async-plan test (asynchronous provisioning and synchronous deprovisioning) - therefore I have kept asynchronous deprovisioning logic.

Thanks for the detailed analysis into https://github.com/mevansam/terraform-provider-cf/pull/52/files/eb4a68f701f66750ca7300816f4f3167b006e016#r194781592

Let's submit an issue "r.service_instance delete perform unnecessary sync delete polling" and record the CLI CLI swallowing of the response code as in a our list of limitations of "client friendly CF CLI api"

In order to have the acceptance tests trigger and report tests passing, please resubmit the PR by pushing your branch onto mevansam/terraform-provider-cf, and please consider squashing the 66 commits as to ease review by the team and avoid polutting the git history.

**Comment by janosbinder** [comment]
Thanks for the feedback @gberche-orange . I have tailored further the opened issue with your comments.

> In order to have the acceptance tests trigger and report tests passing, please resubmit the PR by pushing your branch onto mevansam/terraform-provider-cf, and please consider squashing the 66 commits as to ease review by the team and avoid polutting the git history.

I am afraid to start the acceptance test automatically, I would need access on this repository to trigger the Travis CI. I am happy to open another PR, once I have that, but I can assure you that I have run the acceptance tests yesterday with the credentials Mevan has provided. Currently Travis CI does not support executing acceptance tests on forks, as this would leak out the credentials.

I agree very much on the git history, and it would pollute the history. For these cases GitHub offers squashing the git history into a single commit upon merge. See this article: https://help.github.com/articles/configuring-pull-request-merges/ Then there will a _single_ commit entry in _dev_ commit history like "Merge pull request #52 from chrismathias/async-provisioning-service-instances" See the terraform's commit history for examples: https://github.com/hashicorp/terraform/commits/master

In summary it is a good idea to squash the commits via GitHub when you merge back my changes.

File: cloudfoundry/resource_cf_service_instance.go [hunk 12]
```diff
@@ -82,9 +92,20 @@ func resourceServiceInstanceCreate(d *schema.ResourceData, meta interface{}) (er
 	if id, err = sm.CreateServiceInstance(name, servicePlan, space, params, tags); err != nil {
 		return err
 	}
-	session.Log.DebugMessage("New Service Instance : %# v", id)
+	stateConf := &resource.StateChangeConf{
+		Pending:      resourceServiceInstancePendingStates,
+		Target:       resourceServiceInstanceSucceesStates,
+		Refresh:      resourceServiceInstanceStateFunc(id, "create", meta),
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		PollInterval: 5 * time.Second,
```
Comment (ArthurHlt) [inline cloudfoundry/resource_cf_service_instance.go]: If I do remember well Cloud Controller is polling every 30 seconds maybe poll every 5 seconds is a bit too much

Comment (gberche-orange) [reply cloudfoundry/resource_cf_service_instance.go]: @ArthurHlt considering #107, wouldn't setting a 30s polling interval delay all deletes by 30s?
@janosbinder proposed setting a 5s min timeout, and a 30s refresh period.

**Comment by janosbinder** [comment]
Closing as the branch for this PR is now on the upstream repository (https://github.com/mevansam/terraform-provider-cf/pull/111).
